### ISSUE TITLE
JavaKit: SwiftObjectHolder

### ISSUE THREAD

**Issue by lhoward** [swiftlang/swift-java#168]
No doubt you don't want to ship JARs.

But I've found this useful and maybe it could be an example somewhere. :) (And you can let me know if I've any retain leaks ;))

```java
public class SwiftObjectHolder {
  public final long _swiftObject;

  public SwiftObjectHolder(long swiftObject) {
    _swiftObject = swiftObject;
  }

  native void releaseSwiftObject();
    
  @Override
  public void finalize() {
    if (_swiftObject != 0) {
      releaseSwiftObject();
    } 
  } 
}   
```

```swift
public extension SwiftObjectHolder {
  convenience init(swiftObject: some AnyObject, environment: JNIEnvironment?) {
    let swiftObject = Unmanaged.passRetained(swiftObject)
    let swiftObjectIntPtr = unsafeBitCast(swiftObject, to: Int.self) // Int32 on 32-bit platforms
    self.init(Int64(swiftObjectIntPtr), environment: environment)
  } 
      
  var _unmanagedSwiftObject: Unmanaged<AnyObject>? {
    guard _swiftObject != 0 else { return nil }
    return unsafeBitCast(Int(_swiftObject), to: Unmanaged<AnyObject>.self)
  }

  var swiftObject: AnyObject? {
    _unmanagedSwiftObject?.takeUnretainedValue()
  }
}

extension SwiftObjectHolder: CustomJavaClassLoader {
  public static func getJavaClassLoader(in environment: JNIEnvironment) throws -> JavaClassLoader! {
    _getFlutterSwiftClassLoader()
  }
}

@JavaImplementation("com.padl.FlutterAndroid.SwiftObjectHolder")
extension SwiftObjectHolder: SwiftObjectHolderNativeMethods {
  @JavaMethod
  public func releaseSwiftObject() {
    _unmanagedSwiftObject?.release()
  }
}
```

**Comment by ktoso** [comment]
> No doubt you don't want to ship JARs.

? We'll absolutely be shipping some libraries as jars eventually.

**Comment by ktoso** [comment]
We should not be using finalizers if we can help it.

Please have a look at SwiftKit and the SwiftArena, this is the way to manage swift resources in Java, it's a more predictable modern way of doing so. There is an `AutoArena` which uses phantom references to release when references are dropped in a more efficient way.

**Comment by lhoward** [comment]
Ah, just thought that would make the build process even more of a pain the proverbial!

**Comment by ktoso** [comment]
Can we rephrase this ticket in terms of the problem to solve rather than propose the solution? 

The "modern way" is `SwiftArena` and we should look into using it first if this is about any plain old "keep object alive while i have references to it in java".

**Comment by lhoward** [comment]
Sure. I wasn't aware of SwiftArena. But it looks like it requires Panorama, whereas Android I believe is still at JDK 17. Will confirm shortly.

**Comment by ktoso** [comment]
So funnily enough SwiftArena is not a JDK `Arena` type, the `AutoSwiftArena` just uses phantom references which have been around in the JDK for long. I think we should pursue making that lib have some parts which are usable on older platforms as well, and that may include the `AutoSwiftArena` perhaps - then we don't have to invent another new way to solve the same issue 🤔

**Comment by ktoso** [comment]
Thanks for the rename! :)

**Comment by lhoward** [comment]
Sounds good. I may not have the cycles to help on this but I'll have a think!

**Comment by lhoward** [comment]
Pleasure. Here I was thinking I had thought of something you hadn't.

**Comment by ktoso** [comment]
I'm sure you'll bump into a lot of things we've not thought about yet, this one I had a sneaky plan for though :)

**Comment by lhoward** [comment]
The biggest remaining issue for me is object/JNIEnv affinity, but there's an open issue for that 😉

**Comment by lhoward** [comment]
> We should not be using finalizers if we can help it.
> 
> 
> 
> Please have a look at SwiftKit and the SwiftArena, this is the way to manage swift resources in Java, it's a more predictable modern way of doing so. There is an `AutoArena` which uses phantom references to release when references are dropped in a more efficient way.

Amusingly SwiftKit uses _typeByName() which I [proposed](https://github.com/swiftlang/swift/issues/42998) nearly ten years ago. Must be getting old.

**Comment by ktoso** [comment]
We will actually stop using that as soon as I finish my “extract build plugin” PR, but yeah it’s still hidden api…

**Comment by lhoward** [comment]
Of potential note: https://github.com/vova7878/PanamaPort

**Comment by ktoso** [comment]
Oh that's very interesting, perhaps a route we could attempt huh 🤔 Thanks for sharing!

**Comment by lhoward** [comment]
I think I will just migrate my code to use Cleanable instead of finalize(). I don't have the time right now to unpick the back-portable bits of SwiftArena.

**Comment by ktoso** [comment]
Yeah that would already be a nice step! And that’ll then be easy to hook into the arena :) I can look at these later on

**Comment by lhoward** [comment]
> Yeah that would already be a nice step! And that’ll then be easy to hook into the arena :) I can look at these later on

Pasted in code/links to (edited) original post in case you feel like taking a look. Perhaps I should add C JNI wrappers for `swift_retain` and `swift_release` rather than using `Unmanaged`.

**Comment by ktoso** [comment]
Yeah that's right.

This is done by https://github.com/swiftlang/swift-java/blob/main/SwiftKit/src/main/java/org/swift/swiftkit/AutoSwiftMemorySession.java#L47-L51 already, no need to reinvent -- just reorganize the code so this part of it lives in a non-FFM dependent part of the project. We can make some `SwiftKitBasic` maybe, and the destroy Runnables actually doing FFM invocations into native we'd then have in a separate module that _does use FFM_. but we can have others -- like this one just reference counting or using JNI for the destroy somehow?

I think we should look into sharing the infrastructure basically

**Comment by lhoward** [comment]
Sounds good. Won't be able to look at this again for a few weeks.

**Comment by ktoso** [comment]
We'll work out something like this with @madsodgaard during his jextract+jni work most probably!

**Comment by ktoso** [comment]
We now have "swift arenas" also in JNI mode, and the AutoSwiftMemorySession handles cleanups without FFM as well; https://github.com/swiftlang/swift-java/blob/main/SwiftKitCore/src/main/java/org/swift/swiftkit/core/AutoSwiftMemorySession.java
